src/lobpcg/algorithm.rs

```diff
@@ -0,0 +1,658 @@
+
+    Ok((u, gram_vv_fac))
+}
+
+/// Eigenvalue solver for large symmetric positive definite (SPD) eigenproblems
+///
+/// # Arguments
+/// * `a` - An operator defining the problem, usually a sparse (sometimes also dense) matrix
+/// multiplication. Also called the "stiffness matrix".
+/// * `x` - Initial approximation of the k eigenvectors. If `a` has shape=(n,n), then `x` should
+/// have shape=(n,k).
+/// * `m` - Preconditioner to `a`, by default the identity matrix. Should approximate the inverse
+/// of `a`.
+/// * `y` - Constraints of (n,size_y), iterations are performed in the orthogonal complement of the
+/// column-space of `y`. It must be full rank.
+/// * `tol` - The tolerance values defines at which point the solver stops the optimization. The approximation
+/// of a eigenvalue stops when then l2-norm of the residual is below this threshold.
+/// * `maxiter` - The maximal number of iterations
+/// * `order` - Whether to solve for the largest or lowest eigenvalues
+///
+/// The function returns an `LobpcgResult` with the eigenvalue/eigenvector and achieved residual norm
+/// for it. All iterations are tracked and the optimal solution returned. In case of an error a
+/// special variant `LobpcgResult::NotConverged` additionally carries the error. This can happen when
+/// the precision of the matrix is too low (switch then from `f32` to `f64` for example).
+pub fn lobpcg<
+    A: Float + NdFloat + Sum + ScalarOperand + PartialOrd + Default,
+    F: Fn(ArrayView2<A>) -> Array2<A>,
+    G: Fn(ArrayViewMut2<A>),
+>(
+    a: F,
+    mut x: Array2<A>,
+    m: G,
+    y: Option<Array2<A>>,
+    tol: f32,
+    maxiter: usize,
+    order: Order,
+) -> LobpcgResult<A> {
+    // the initital approximation should be maximal square
+    // n is the dimensionality of the problem
+    let (n, size_x) = (x.nrows(), x.ncols());
+    assert!(size_x <= n);
+
+    /*let size_y = match y {
+        Some(ref y) => y.ncols(),
+        _ => 0,
+    };
+
+    if (n - size_y) < 5 * size_x {
+        panic!("Please use a different approach, the LOBPCG method only supports the calculation of a couple of eigenvectors!");
+    }*/
+
+    // cap the number of iteration
+    let mut iter = usize::min(n * 10, maxiter);
+    let tol = NumCast::from(tol).unwrap();
+
+    // calculate cholesky factorization of YY' and apply constraints to initial guess
+    let cholesky_yy = y.as_ref().map(|y| {
+        let cholesky_yy = y.t().dot(y).cholesky().unwrap();
+        apply_constraints(x.view_mut(), &cholesky_yy, y.view());
+        cholesky_yy
+    });
+
+    // orthonormalize the initial guess
+    let (x, _) = match orthonormalize(x) {
+        Ok(x) => x,
+        Err(err) => return LobpcgResult::NoResult(err),
+    };
+
+    // calculate AX and XAX for Rayleigh quotient
+    let ax = a(x.view());
+    let xax = x.t().dot(&ax);
+
+    // perform eigenvalue decomposition of XAX
+    let (mut lambda, eig_block) = match sorted_eig(xax.view(), None, size_x, &order) {
```

Comment:
You can pass in `xax` by value instead of by view, since it's not used afterwards.